elastic/elasticsearch
# Integrate IBM watsonx to Inference API for text embeddings

This PR adds [IBM watsonx AI](https://dataplatform.cloud.ibm.com/docs/content/wsj/analyze-data/fm-embed-overview.html?context=wx) text embeddings support for the [inference APIs](https://www.elastic.co/guide/en/elasticsearch/reference/current/inference-apis.html).

### Prerequisites to Model Creation
- An IBM Account with reservation to watsonx AI. Please read this [doc](https://docs.google.com/document/d/1_zn2Qz_CPJJHnma5BNQ4lbVOh-bvi55LKGqubd-SYTg/edit#heading=h.ivczzpjehg2u) on how to get access.
- An API key used to access IBM cloud

### Inference Model Creation:

```
PUT _inference/text_embedding/{inference_model_id}
{
   "service": "watsonxai",
    "service_settings": {
        "api_key": "<<API_KEY>>",
        "url": "<<URL_TO_IBM_CLOUD>>",
        "model_id": "<<MODEL_ID>>",
        "project_id": "<<PROJECT_ID>>",
        "api_version": "<<API_VERSION>>"
    }
}
```

### Inference Model Inference
```
POST _inference/text_embedding/{inference_model_id}
{
  "input": <single string, or array of strings>
}
```

# Manual test
Following steps have been taken to test watsonx ai integration
- Create an inference endpoint using watsonx ai
- Generate text_embeddings using the inference endpoint
- Create an index
- Create a semantic_text field for the new index and refer to watson ai inference endpoint
- Verify that embeddings are generated for the semantic_text field while inserting data to the index

Please find the testing in the following screen-recording:

https://github.com/user-attachments/assets/38cc0573-644d-49e9-b184-97f5676c4f71
